src/cosim/utility/thread_pool.hpp

```diff
@@ -29,30 +29,28 @@ class thread_pool
     std::queue<std::function<void()>> work_queue_;
     std::vector<std::thread> threads_;
     std::mutex m_;
-    std::condition_variable cv_;
+    std::condition_variable cv_finished_;
+    std::condition_variable cv_worker_;
     unsigned int pending_tasks_;
 
     void worker_thread()
     {
+        std::chrono::seconds wait_duration(1);
         while (!done_) {
-
             std::unique_lock<std::mutex> lck(m_);
             if (!work_queue_.empty()) {
-
                 auto task = std::move(work_queue_.front());
                 work_queue_.pop();
-                lck.unlock();
 
+                lck.unlock();
                 task();
 
                 lck.lock();
                 pending_tasks_--;
                 lck.unlock();
-
-                cv_.notify_one();
-
+                cv_finished_.notify_one();
             } else {
-                std::this_thread::yield();
+                cv_worker_.wait_for(lck, wait_duration);
```

Comment:
I don't really agree with this. Can't we just block the thread until work is available through submit()? This looks like it'll wake the thread up once every second, but I don't immediately see a reason to do so.

Reply:
That sounds like a synchronization issue? 

I think we need to use the mutex in the destructor also; 

```
std::unique_lock<std::mutex> lck(m_);
done_ = true;
cv_worker_.notify_all();
lck.unlock();
```

Reply:
From what I understand, `notify_all` can be missed by the worker thread if it is issued while the thread is not in the waiting state (i.e. not blocked by `wait()`). And since `notify_all` can be issued at any arbitrary time (e.g. when the program terminates), it is difficult to control. So it would be safer and cleaner for the thread just to 'timeout' periodically with a reasonable interval to check if `done_=true`.

And yes I think we can guard `done_` in the destructor.

Reply:
I committed some changes now, mainly to facilitate the following suggestions: 

* If no work is available, immediately block the worker thread via `cv_worker_`. Revive when we're either `done_`, or `work_queue_ `is not empty (don through `notify_one()` in `submit()`). 
* `wait_for_tasks_to_finish` should wait `pending_tasks_ `and the `work_queue_ ` to empty, not the threads (!)
* Increment pending_tasks_ only when the worker thread "begins" (may not make much of a difference, but seems cleaner) 
* Lock the destructor also via the mutex in order to modify `done_ `and call `notify_all()`. Since all modifications of `done_ `are now in a mutex, it can also be changed to a regular `bool`. In addition I removed the `joinable()` query - if all threads are either running or notified, do we need it? 

I left the previous code in comments for easier comparison. Please have a look.

Reply:
I tested the updated version with several test runs, and seems the program terminates correctly (without blocking).
Some comments:
1. Wondering if the worker thread can miss `notify_all` when it is issued while the thread is just before acquiring the lock? (`std::unique_lock<std::mutex> lck(m_)`).
2. The lock does not go out of scope at the end of `if` block unlike the comment mentioned.

Reply:
According to the `condition_variable `documentation, it's not necessary to keep the lock when calling `notify_one `or `notify_all`, so I've changed that back to manually calling `unlock()` in both cases.

Reply:
Further, the documentation details how `condition_variable `should be used in tandem with a `mutex`, both when modifying a shared variable, and waiting on the condition:

https://en.cppreference.com/w/cpp/thread/condition_variable

I'd say the code is correct now based on this, I can't really spot any deviations from these requirements.

Reply:
Seems `wait()` with the pred will run the pred before the thread is put into wait https://en.cppreference.com/w/cpp/thread/condition_variable/wait. Since done_ is guarded, I think this implementation is fine :)